File: 2019/day-3/python/crossed_wires.py
import csv
import numpy as np

def compute_wire_segments(wires):
    """ Given the instructions in wires, the function returns an (N+1)x4 array
    of wire positions. First two elements are the x,y positions, the third
    element is the cummulated wire length, and the fourth element indicates if
    the line is vertical or horizontal."""

    # Wire segments container
    wire_segments = []

    # Go through each wire
    for wire in wires:

        # Start position
        pos = np.zeros((1,4))

        # Initialize array
        wire_segment = np.zeros((len(wire)+1,4))

        # Move 'along' wire
        for i, w in enumerate(wire):

            # Initialize direction steps and orientation
            dx,dy,step,orientation = 0,0,int(w[1:]),0

            # Wire moves left or right
            if w[0] == 'L' or w[0] == 'R':
                dx = step if (w[0] == 'R') else -step
            # Wire move up or down
            if w[0] == 'U' or w[0] == 'D':
                dy = step if (w[0] == 'U') else -step
                orientation = 1

            # Update position
            pos += [dx,dy,step,orientation]

            # Add position to wire_segment
            wire_segment[i+1,:] = pos

        # Add wire segment to container
        wire_segments.append(wire_segment)

    # Return segments
    return wire_segments

def intersect(line1,line2):
    """ Checks if line1: [A,B] intersects with line2: [C,D]. Assumes only
    horizontal and vertical lines."""

    # Check if lines are perpendicular
    if line1[1,-1] == line2[1,-1]:
        return None,None

    # Compute min and max bounds for line 1
    x1min,x1max = min(line1[:,0]),max(line1[:,0])
    y1min,y1max = min(line1[:,1]),max(line1[:,1])

    # Compute min and max bounds for line 2
    x2min,x2max = min(line2[:,0]),max(line2[:,0])
    y2min,y2max = min(line2[:,1]),max(line2[:,1])

    # Intersection condition
    if (x1max >= x2min) and (x1min <= x2max) and \
        (y1max >= y2min) and (y1min <= y2max):

        # Compute intersection point
        p = [x2min,y1min] if (x1max-x1min) != 0 else [x1min,y2min]

        # Total wire steps
        wire_steps = line1[0,2] + line2[0,2]

        # Added wire steps to intersection point
        steps_to_point = sum(abs(line1[0,:2] - p)) + sum(abs(line2[0,:2]-p))

        return p,(wire_steps+steps_to_point)

    return None,None

def intersect_wires(wire1,wire2):
    """ Loop through the wire segments wire1 and wire2 and looks for possible
    intersections."""

    intersection, steps = [], []
    for i in range(len(wire1)-1):
        for j in range(len(wire2)-1):
            point,s = intersect(wire1[i:i+2],wire2[j:j+2])
            if point is not None:
                intersection.append(point)
                steps.append(s)
    return intersection,steps

def crossed_wires(wires):
    """ This function assumes that the wires aren't routed along eachother, i.e.,
	they will only cross perpendicularly, hence it is enough only to check
    vertical wire segments vs horizontal wire segments."""

    # Compute the wire positions for each wire
    wire_segments = compute_wire_segments(wires)

    # Try to intersect the two wire segments with eachother
    points,steps = intersect_wires(wire_segments[0],wire_segments[1])

    # Compute Manhattan distance
    minimum_manhattan_distance = int(min(np.sum(np.abs(points),axis=1)))

    # Compute fewest steps
    fewest_steps  = int(min(steps))

    return minimum_manhattan_distance,fewest_steps

def main():
  # Open data file and read through all lines
  file_location = "../data.csv"
  try:
    wires = []
    with open(file_location) as csvfile:
      reader = csv.reader(csvfile,delimiter=',')
      # Go through each row of the csv
      for row in reader:
          wires.append(row)

    manhattan_distance,fewest_steps = crossed_wires(wires)
    print("Solution to part one: " + str(manhattan_distance))
    print("Solution to part two: " + str(fewest_steps))

  except IOError:
    print("Cannot find file at: " + file_location)

if __name__ == "__main__":
  main()

File: 2019/day-3/cpp/src/crossed_wires.cpp
#include "crossed_wires.hpp"

using std::vector;
using std::string;

CrossedWires::CrossedWires(string fileName) : wire{}, manhattan_distance{0}, n_steps{0} {
    if(readFile(fileName,wire)){
      crossed_wires(wire,manhattan_distance,n_steps);
    }
}

void CrossedWires::crossed_wires(vector<vector<string>> & input, int & dist, int & step) {

  vector<vector<int>> wire_segments[2];
  compute_wire_segments(input,wire_segments);

  vector<int> points;
  vector<int> steps;
  intersect_wires(wire_segments[0],wire_segments[1],points,steps);
  dist = *std::min_element(points.begin(),points.end());
  step = *std::min_element(steps.begin(),steps.end());

}

void CrossedWires::compute_wire_segments(vector<vector<string>> & wires,vector<vector<int>> wire_segments[2]) {

  for (unsigned int i = 0; i < wires.size(); i++) {
    vector<int> pos {0, 0, 0, 0};
    vector<vector<int>> wire_segment; wire_segment.push_back(pos);

    for (unsigned int j = 0; j < wires[i].size(); j++) {
      string w = wires[i][j];
      int dx = 0, dy = 0, orientation = 0;
      int step = std::stoi(w.substr(1,w.back()));

      if (w[0] == 'L' || w[0] == 'R')
        dx = (w[0] == 'R') ? step : -step;
      else if (w[0] == 'U' || w[0] == 'D') {
        dy = (w[0] == 'U') ? step : -step;
        orientation = 1;
      }
      pos[0]+=dx, pos[1]+=dy, pos[2]+= step, pos[3]=orientation;
      wire_segment.push_back(pos);
    }
    wire_segments[i] = wire_segment;
  }
}

void CrossedWires::intersect_wires(vector<vector<int>> & wire1,vector<vector<int>> & wire2, vector<int> & dist, vector<int> & steps) {

  for (unsigned int i = 0; i < wire1.size()-1; i++){
    for (unsigned int j = 0; j <wire2.size()-1; j++) {
      vector<int> point; int st;
      if(intersect(wire1[i],wire1[i+1],wire2[j],wire2[j+1],point,st)){
        dist.push_back(std::abs(point[0])+std::abs(point[1]));
        steps.push_back(st);
      }
    }
  }
}

bool CrossedWires::intersect(vector<int> & A, vector<int> & B,vector<int> & C, vector<int> & D,vector<int> & point,int  & steps) {


  if (B[3]==D[3]){
    return false;
  }
  int x1min = std::min(A[0],B[0]), x1max = std::max(A[0],B[0]);
  int y1min = std::min(A[1],B[1]), y1max = std::max(A[1],B[1]);

  int x2min = std::min(C[0],D[0]), x2max = std::max(C[0],D[0]);
  int y2min = std::min(C[1],D[1]), y2max = std::max(C[1],D[1]);

  if ((x1max >= x2min) && (x1min <= x2max) && (y1max >= y2min) && (y1min <= y2max)){
    // Compute intersection point
    point = ((x1max-x1min) != 0) ? vector<int>{x2min,y1min} : vector<int>{x1min,y2min};
    int wire_steps = A[2] + C[2];

    int steps_to_point = std::abs(A[0]-point[0]) + std::abs(A[1]-point[1]) +
                        std::abs(C[0]-point[0]) + std::abs(C[1]-point[1]);
    steps = wire_steps+steps_to_point;

    return true;
  }
  return false;
}

bool CrossedWires::readFile(string fileName, vector<vector<string>> & output) {
  std::ifstream file(fileName);
  if(!file) {
    std::cerr << "Cannot open the File : " << fileName << std::endl;
    return false;
  }
  std::string row, instruction;
  while(std::getline(file,row)) {
    std::stringstream ss(row);
    std::vector<std::string> results;
    while (ss.good()) {
      getline(ss,instruction,',');
      results.push_back(instruction);
    }
    output.push_back(results);
  }
  file.close();
  return true;
}
